#28 graph6 adapter

pictavien opened the pull request:
Hi,

I have implemented a Graph6/Digraph6 TopologySerializer. 
The Sparse6 format is not supported.

remikey commented:
Thank you for the contribution @pictavien . Well done as usual! Some remarks after our discussion.

* `BitVector.java`
  * please replace magic numbers by constants

* `Codec.java`
  * Regarding our class naming discussion:
    * something like `Graph6Codec.java` or `Graph6BVHelper.java`?
    * rename `Graph6.java` into `Graph6Encoder.java`?
  * In order to ease readability, although the class is package private, I suggest that the main methods be public:
    * `Codec.N`
    * `Codec.R`
    * `Codec.readN`
    * `Codec.readR`

* `Graph6.java`
  * Rename into `Graph6Encoder.java`?
  * Please provide a class comment
  * I think that the commented `tp.setOrientation()` call should be un-commented
  * In order to ease readability, although the class is package private, I suggest that the main methods be public:
    * `encodeGraph`
    * `decodeGraph`

* `BitVectorTest.java`
  * To increase readability, please make a `void assertEquals(boolean[] expected, BitVector actual, int test_index)` method for last part which tests the whole BitVector consistency

* `CodecTest.java`
  * Rename according to production class renaming
  * Unused imports are present
  * Unused fields are present
  * Please add a comment for data provenance
  * Unused parameters are present due to Parametrized tests.

    Something like this should do the job
    ```
    void checkREncodingPaddedStrings(String bits,  byte[] expectedBV,
                                     @SuppressWarnings("unused") byte[] expectedR) {
    ```

* `Graph6Test.java`
  * Rename according to production class renaming
  * Please replace utility code blocks by a method call (e.g. createTopology(), loadInputFiles())
  * Unused variables are present
  * We are using a deprecated version of `getLink()`
  * parameter `G` does not match Java naming conventions

* `Graph6TopologySerializerTest.java`
  * It is a bit strange that this file does not exist

* test resources
  * Create a readme file to list the provenance of the resources
  * `digraphs.g6` only contains one example when `graphs.g6` has loads of them
  * Please add resource files and tests for graphs with headers (this case is currently only tested by encoding/decoding it ourselves)

* `JViewer.java` / `JBotSimConvert.java` / `JBotSimViewer.java`
  * create `Graph6TopologySerializer.GRAPH6_FILENAME_EXTENSIONS` and use it as it is done for `DotTopologySerializer`
  * We could use the same object twice by creating a variable ( `DotTopologySerializer` creation could be improved also)
  

* Remark to be adressed in futur versions:
  * `Graph6TopologySerializer.java`: should we disable wireless capabilities on the Topology?
  * `Graph6TopologySerializer.java`: we could propose an option to reorganize the Topology spatialy according to WIRED links
  * `TopologySerializerFilenameMatcher.java`: centralized the code used in `JViewer.java`, `JBotSimConvert.java`, `JBotSimViewer.java` in a method called `initializeDefaultSerializers()`?

pictavien commented:
>     * `Codec.java`      
>       * Regarding our class naming discussion:         
>         * something like `Graph6Codec.java` or `Graph6BVHelper.java`?
I've chosen  `Graph6BVHelper` (the reason is below).
>       * In order to ease readability, although the class is package private, I suggest that the main methods be public:
Ok. I applied the advice (also in other classes) but I maintain the package visibility for members that are used only by tests.
         
>     * `Graph6.java`       
>       * Rename into `Graph6Encoder.java`? 
I've renamed the class as `Graph6Codec` because it is used for both usages.
> 
>     * `BitVectorTest.java`
>       
>       * To increase readability, please make a `void assertEquals(boolean[] expected, BitVector actual, int test_index)` method for last part which tests the whole BitVector consistency

Done. I also transform the test as a parameterized test to ease the debugging.
 
>     * `Graph6Test.java`
>     * `Graph6TopologySerializerTest.java`
As discussed, I've move tests related to graphs stored in resource files in a new class `Graph6TopologySerializerTest`.  I also add a test for headers.
 
>     * test resources
>       * `digraphs.g6` only contains one example when `graphs.g6` has loads of them
I've not yet found a database with such graphs. All directed graphs I've found, was in an other format.
>       * Please add resource files and tests for graphs with headers (this case is currently only tested by encoding/decoding it ourselves)
Done. I've also reduced the number of input graphs (basically `graphs.g6` contained 18K graphs). I've randomly selected 10% if graphs from the original set.

>     * `JViewer.java` / `JBotSimConvert.java` / `JBotSimViewer.java`      
>       * create `Graph6TopologySerializer.GRAPH6_FILENAME_EXTENSIONS` and use it as it is done for `DotTopologySerializer`
>       * We could use the same object twice by creating a variable ( `DotTopologySerializer` creation could be improved also)
Both are done. I've also added a new helper method in FilenameMatcher to register an array of 
extensions. 

>     * Remark to be adressed in futur versions:       
>       * `Graph6TopologySerializer.java`: should we disable wireless capabilities on the Topology?
>       * `Graph6TopologySerializer.java`: we could propose an option to reorganize the Topology spatialy according to WIRED links
>       * `TopologySerializerFilenameMatcher.java`: centralize the code used in `JViewer.java`, `JBotSimConvert.java`, `JBotSimViewer.java` in a method called `initializeDefaultSerializers()`?

remikey commented:
ok for me.
